tailwarden/komiser
# fix: improve backend testsuite

## Problem

The testsuite in CI/CD and localy currently reports
```logtalk
Run make test
go test -v 
?   	github.com/tailwarden/komiser	[no test files]
```

## Solution

Tests are executed like `go test -v $(go list ./... | grep -v /dashboard/)` and this is suppossed to work fine, however for unknown reasons it does not when executed through the `Makefile`. Updating the command to `go test ./... | grep -v /dashboard/` and pipe the ignore filter instead works more reliably

## Changes Made

- Update `Makefile` to execute all found tests except inside `/dashboard` folder
- Update GCP pricing map (**[SEE NOTES ](#notes)**)
- Update `CONTRIBUTING.md`
- Update tests `machine_test.go` and refactor how to test because previously we were mapping against a fixed price. I presume (actually it's hope) that the schema changes less often than the prices. I made the tests more dry and only check if the price returned is above a certain number. The number is currently set to `0`.
- Update Feedback Widget to use new Toast component

## How to Test

- Localy: checkout this branch and run `make test` or `go test ./... | grep -v /dashboard/`
- CI/CD: check the run results of the latest commit in this PR. Is should look similar to the attached screenshot!

## Screenshots

![Screenshot 2023-10-20 at 13 47 34](https://github.com/tailwarden/komiser/assets/320272/d1e4cee0-4e19-489f-b3de-7d30dbeeac48)
![Screenshot 2023-10-20 at 13 25 03](https://github.com/tailwarden/komiser/assets/320272/19f8d6d4-6ec0-4595-b62e-bd4eb5f367dc)


## Notes

There are test fixes in here as well.
For example `/utils/gcpcomputepricing` was broken. It seems like Google updated their pricing structure? 
Comparing our current `VmsOnDemandMemoryPerGb` type with the content returned from [gcp-compute.json](https://www.gstatic.com/cloud-site-ux/pricing/data/gcp-compute.json) it seems they now nest `Vmimagee2RAME2` instance type under `E2` (sdee screenshot attached). 
I updated the types being used in the utils test and the test now passes without error.

Before the tests where returning

```logtalk
--- FAIL: TestGet (0.32s)
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0
--- FAIL: TestGetE2standard32OnDemand (0.08s)
    machine_test.go:62: core price not found for "us-west1" region
```

Also check comment thread please.

## Checklist


## Reviewers

@[username of the reviewer]
